### PR TITLE
Bulk scope test recorder

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
1. Add support  for new scope facets to a particular language. eg: `scopeSupportFacets\typescript.ts`
2. Issue first command to show all unimplemented facets 
   * Select language
   * New untitled document opens with all `supported` facets for the language that are missing `.scope` files
3. Edit document with any number of fixtures
4. Issues second command to create multiple scope test fixtures

```
[[typescript]]

[command] - A command, for example Talon spoken command or bash
hello
---
```

Thoughts:
* Do we want to close the document after it's read? If yes do we have something on the ide for this already or do I need to add it?
* I'm not one hundred percent sure on the command identifiers
* Regarding docs. Should I add a new heading in the exist [test case recorder file](https://github.com/cursorless-dev/cursorless/blob/ec694b754cd7169f792e6d2bb4028ab17a83bf4e/docs/contributing/test-case-recorder.md) or should I start a new one for this format?

## Checklist

- [/] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [ ] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [/] I have not broken the cheatsheet

### CODE REVIEW

**Comment by AndreasArvidsson** [comment]
@pokey Scope faced description is now added

**Comment by pokey** [comment]
Any idea how we would go about recording tests for eg `javascript.core`? I guess for now just record them for javascript and then move them?

**Comment by AndreasArvidsson** [comment]
> Any idea how we would go about recording tests for eg `javascript.core`? I guess for now just record them for javascript and then move them?

That was more or less what I was thinking

**Comment by pokey** [comment]
yeah probably fine for now. I don't see any easy way to do something smarter, and I believe for everything other than ts/js/tsx/jsx you could get away with just recording the test for the base language. Eg for css/scss you'd just add the support info to css and record test for css and it would just work

**Comment by AndreasArvidsson** [comment]
Pretty much.

**Review by pokey — COMMENTED** [review]
Ok I made [some tweaks](https://github.com/cursorless-dev/cursorless/pull/2383/files/26b62e97601df779076701a7c57c106ad95855fb..75aec1af24179c7228411057a6595cf1ccc772cd) and left a couple comments. My changes were as follows:

- Add spoken forms to our dev talon file
- Use a snippet so that you can just say "next" to go from one facet to the next
- Set languade id to markdown, as the highlighting is not bad and it prevents VSCode from trying to guess a language
- Fix the regex for scopes that have a `.` in them

File: packages/common/src/scopeSupportFacets/languageScopeSupport.ts [hunk 1]
```diff
@@ -29,7 +29,7 @@ import { yamlScopeSupport } from "./yaml";
 
 export const languageScopeSupport: Record<
   string,
-  LanguageScopeSupportFacetMap
+  LanguageScopeSupportFacetMap | undefined
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/languageScopeSupport.ts:32]: The key is a string so I think this is much clearer that you can actually get back undefined. I actually think the implementation of record in typescript is a bit problematic where it does not have the key value map expectation where the key can be missing.

File: changelog/2024-06-bulkScopeTestRecorder.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+tags: [enhancement]
+pullRequest: 2383
+---
+
+- Added bulk scope test recorder commands
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply changelog/2024-06-bulkScopeTestRecorder.md]: You are right. File remove.

**Comment by AndreasArvidsson** [comment]
I'm happy with those changes

File: packages/common/src/scopeSupportFacets/languageScopeSupport.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (auscompgeek) [reply packages/common/src/scopeSupportFacets/languageScopeSupport.ts:32]: If the key is a literal union then TypeScript will enforce exhaustiveness of the record.

[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/languageScopeSupport.ts:32]: In this case the key type is `string`. If I do `languageScopeSupport["hello"]` I will get back undefined and I think that the type system should reflect this.

[the remark this answers is not in the document]
Comment (pokey) [reply packages/common/src/scopeSupportFacets/languageScopeSupport.ts:32]: Yeah this one has bitten me before. I'd be tempted to have a utility type


```
type StringRecord<T> = Partial<Record<string, T>>
```

[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/languageScopeSupport.ts:32]: Sounds good

[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/languageScopeSupport.ts:32]: Should I implement this here?

[the remark this answers is not in the document]
Comment (pokey) [reply packages/common/src/scopeSupportFacets/languageScopeSupport.ts:32]: Yeah I think so

[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/common/src/scopeSupportFacets/languageScopeSupport.ts]: done

**Review by pokey — APPROVED** [review]

**Pull request merged**